### PR TITLE
criu/plugin: Restore libhsakmt shared memory files

### PR DESCRIPTION
**Pull request by dayatsin-amd**
Libhsakmt(thunk) uses a shared memory file in /dev/shm/hsakmt_shared_mem
and its semaphore in /dev/shm/hsakmt_shared_mem. Adding a check during
checkpoint to see if these two files exist. If they exist then the
plugin will try to restore them during restore.

### CODE REVIEW

File: plugins/amdgpu/amdgpu_plugin.c [hunk 1]
```diff
@@ -621,6 +626,83 @@ void *dump_bo_contents(void *_thread_data)
...
+		fd = shm_open(HSAKMT_SHM, O_CREAT | O_RDWR, 0666);
+
+		ret = ftruncate(fd, shared_mem_size);
+		if (ret < 0) {
+			pr_err("amdgpu_plugin: Failed to truncate shared mem %s\n", HSAKMT_SHM);
+			close(fd);
+			return -EINVAL;
+		}
+
+		shared_mem_ptr = mmap(0, shared_mem_size, PROT_WRITE, MAP_SHARED, fd, 0);
```
Comment (rajbhar) [inline plugins/amdgpu/amdgpu_plugin.c:690]: PROT_READ | PROT_WRITE

File: plugins/amdgpu/amdgpu_plugin.c [hunk 2]
```diff
@@ -621,6 +626,83 @@ void *dump_bo_contents(void *_thread_data)
...
+	if (shared_mem_ptr == MAP_FAILED) {
+		pr_perror("amdgpu_plugin: Failed to mmap shared mem %s\n", HSAKMT_SHM);
+		return -EACCES;
+	}
+
+	*shared_mem_size = st.st_size;
+
+	/* First 4 bytes of shared file is the magic */
+	memcpy(shared_mem_magic, shared_mem_ptr, sizeof(*shared_mem_magic));
+	pr_debug("amdgpu_plugin: Shared mem magic:0x%x\n", *shared_mem_magic);
```
Comment (rajbhar) [inline plugins/amdgpu/amdgpu_plugin.c]: lets use plugin_log_msg since thats compiled only with DDEBUG

**Review by rajbhar — CHANGES_REQUESTED** [review]
looks good overall, some changes requested.

File: plugins/amdgpu/amdgpu_plugin.c (hunk 1 above)
Comment (dayatsin-amd) [reply plugins/amdgpu/amdgpu_plugin.c:690]: I am not reading to this memory. I am just writing

File: plugins/amdgpu/amdgpu_plugin.c (hunk 2 above)
Comment (dayatsin-amd) [reply plugins/amdgpu/amdgpu_plugin.c]: ACK

File: plugins/amdgpu/amdgpu_plugin.c (hunk 1 above)
Comment (rajbhar) [reply plugins/amdgpu/amdgpu_plugin.c:690]: Ok, I was thinking about the global shared table that thunk creates. Then you should munmap in both save and restore calls.

File: plugins/amdgpu/amdgpu_plugin.c [hunk 3]
```diff
@@ -621,6 +626,83 @@ void *dump_bo_contents(void *_thread_data)
...
+	int ret, fd;
+	struct stat st;
+	uint32_t *shared_mem_ptr;
+	sem_t *sem = SEM_FAILED;
+
+	if (!shared_mem_size)
+		return 0;
+
+	if (!stat(HSAKMT_SHM_PATH, &st)) {
+		pr_debug("amdgpu_plugin: %s already exists\n", HSAKMT_SHM_PATH);
```
Comment (rajbhar) [inline plugins/amdgpu/amdgpu_plugin.c:677]: I think you forgot to change this one to plugin_log_msg

Comment (dayatsin-amd) [reply plugins/amdgpu/amdgpu_plugin.c:677]: For this one, it would be better to print if the user calls criu with -vvv

File: plugins/amdgpu/amdgpu_plugin.c [hunk 4]
```diff
@@ -621,6 +626,83 @@ void *dump_bo_contents(void *_thread_data)
...
+			return -EACCES;
+		}
+		memcpy(shared_mem_ptr, &shared_mem_magic, sizeof(shared_mem_magic));
+		close(fd);
+	}
+
+	sem = sem_open(HSAKMT_SEM, O_CREAT, 0666, 0);
+	if (sem == SEM_FAILED) {
+		pr_perror("Failed to create %s\n", HSAKMT_SEM);
+		return -EACCES;
```
Comment (rajbhar) [inline plugins/amdgpu/amdgpu_plugin.c:704]: It returns SEM_FAILED with set error so just return -1 as pr_perror takes care of showing actual error

Comment (dayatsin-amd) [reply plugins/amdgpu/amdgpu_plugin.c:704]: This will not affect behavior. Caller just checks for non-zero.

File: plugins/amdgpu/amdgpu_plugin.c (hunk 1 above)
Comment (dayatsin-amd) [reply plugins/amdgpu/amdgpu_plugin.c:690]: ACK for munmap

File: plugins/amdgpu/amdgpu_plugin.c (hunk 4 above)
Comment (rajbhar) [reply plugins/amdgpu/amdgpu_plugin.c:704]: This is just a nit and I can take care of this while applying this PR.  @fxkamd if you don't have any objections to this commit, I think its good to get included in our staging.

File: plugins/amdgpu/amdgpu_plugin.c [hunk 5]
```diff
@@ -621,6 +626,86 @@ void *dump_bo_contents(void *_thread_data)
...
+	fd = shm_open(HSAKMT_SHM, O_RDONLY, 0);
+	if (fd < 0) {
+		*shared_mem_size = 0;
+		return 0;
+	}
+
+	shared_mem_ptr = mmap(0, st.st_size, PROT_READ, MAP_SHARED, fd, 0);
+	if (shared_mem_ptr == MAP_FAILED) {
+		pr_perror("amdgpu_plugin: Failed to mmap shared mem %s\n", HSAKMT_SHM);
+		return -EACCES;
```
Comment (fxkamd) [inline plugins/amdgpu/amdgpu_plugin.c:651]: You should close the fd here.

File: plugins/amdgpu/amdgpu_plugin.c (hunk 4 above)
Comment (fxkamd) [reply plugins/amdgpu/amdgpu_plugin.c:704]: We're making up error codes in a number of places. If we wanted to return the real error code, we'd return -errno. But the caller ignores this anyway. So it really doesn't matter.

File: plugins/amdgpu/Makefile [hunk 6]
```diff
@@ -9,14 +9,15 @@ include $(__nmk_dir)msg.mk
 
 CC      		:= gcc
 PLUGIN_CFLAGS  		:= -g -Wall -Werror -D _GNU_SOURCE -shared -nostartfiles -fPIC
+PLUGIN_LDFLAGS		:= -lrt -lpthread
```
Comment (fxkamd) [inline plugins/amdgpu/Makefile:12]: This seems unrelated to the shared memory changes. Should this be part of the commit that adds multi-threaded saving of VRAM?

Comment (dayatsin-amd) [reply plugins/amdgpu/Makefile:12]: the shm_* functions need -lrt, and the sem_* functions need -lpthread.
The multi-threaded saving of VRAM should need:
+PLUGIN_LDFLAGS		:= -lpthread

I will add that to the multi-threaded commit if it is needed.

File: plugins/amdgpu/amdgpu_plugin.c (hunk 5 above)
Comment (dayatsin-amd) [reply plugins/amdgpu/amdgpu_plugin.c:651]: ACK

**Comment by dayatsin-amd** [comment]
Commit included in 
https://github.com/RadeonOpenCompute/criu-new/pull/5
